Bluetooth: Classic: add power mode control with sniff mode

chengkai15 opened the pull request:
# Bluetooth: Classic: add power mode control with sniff mode

## Summary
This PR implements power mode control for Bluetooth Classic connections, adding support for sniff mode to enable power saving while maintaining connectivity. The feature includes comprehensive APIs, event handling, and shell commands for testing and configuration.

## Changes
- **Power Mode APIs**: Add `bt_conn_br_enter_sniff_mode()` and `bt_conn_br_exit_sniff_mode()` APIs with parameter validation
- **Mode Change Notification**: Implement HCI mode change event handling and `br_mode_changed` callback for application notification
- **Shell Integration**: Add `set_sniff_mode` shell command for testing power mode transitions
- **Configuration Option**: Introduce `CONFIG_BT_POWER_MODE_CONTROL` Kconfig option to enable the feature

## Technical Details
- Supports configurable sniff mode parameters: min/max interval, attempt count, and timeout
- Handles mode transitions between active and sniff modes with proper state tracking
- Implements comprehensive parameter validation according to Bluetooth specification
- Provides real-time callback notifications for mode changes including interval information
- Maintains connection state consistency during power mode transitions

## Features
- **Fine-grained Control**: Applications can optimize power consumption based on use case requirements
- **Event-driven Notifications**: Applications receive callbacks when connection mode changes
- **Shell Testing**: Complete shell integration for development and testing purposes
- **Backward Compatibility**: Maintains existing functionality when feature is disabled

## Testing
- Verified sniff mode entry/exit functionality via shell commands
- Tested parameter validation with various valid and invalid inputs
- Confirmed mode change callback notifications work correctly
- Validated state consistency during mode transitions
- Tested interoperability with remote Bluetooth devices

## Impact
- Enables significant power savings for Bluetooth Classic applications
- Provides foundation for advanced power management features
- Maintains full Bluetooth specification compliance
- No impact on existing functionality when disabled via Kconfig

This power mode control feature is essential for battery-powered Bluetooth Classic devices requiring optimized power consumption while maintaining reliable connections.

gzh-terry commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -1987,6 +1987,18 @@ struct bt_conn_cb {
+#if defined(CONFIG_BT_POWER_MODE_CONTROL)
+	/** @brief The connection mode change
+	 *
+	 *  This callback notifies the application controller mode changed
```
Alternatively, change the description of mode to "Active/Hold/Sniff mode." if this is a "HCI mode change event"

jhedberg commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -1987,6 +1987,18 @@ struct bt_conn_cb {
+	 *  @param mode Active/Sniff mode.
+	 *  @param interval Sniff interval.
+	 */
+	void (*mode_changed)(struct bt_conn *conn, uint8_t mode, uint16_t interval);
```
Since this is limited to Bluetooth Classic, but it's inside a generic callback structure (`struct bt_conn`), would it make sense for the callback to have a `br_*` prefix?

chengkai15 committed:
Bluetooth: Classic: add power mode control for sniff mode

Implement BR/EDR power mode control with sniff mode functionality.
Adds APIs bt_conn_br_enter_sniff_mode() and bt_conn_br_exit_sniff_mode()
to manage power saving modes. Includes parameter validation and HCI
command handling for sniff mode configuration with min/max intervals,
attempt count, and timeout parameters.

Signed-off-by: Kai Cheng <chengkai@xiaomi.com>

chengkai15 committed:
Bluetooth: Classic: add mode change notification for sniff mode

Implement mode change event handling and callback notification for
BR/EDR power mode transitions. Adds br_mode_changed callback to
notify applications when connection switches between active and
sniff modes. Handles HCI mode change events and propagates mode
and interval information to registered callbacks.

Signed-off-by: Kai Cheng <chengkai@xiaomi.com>

chengkai15 committed:
Bluetooth: shell: add sniff mode control command

Add shell command for testing BR/EDR power mode control. Supports
entering sniff mode with configurable parameters (min/max interval,
attempt, timeout) and exiting back to active mode. Provides real-time
feedback on mode change requests and status.

Signed-off-by: Kai Cheng <chengkai@xiaomi.com>

gzh-terry commented on subsys/bluetooth/host/conn.c:
Use 	`if (!bt_conn_is_type(conn, BT_CONN_TYPE_BR))`

Which includes a `null` check for `conn`.

```
	if (!bt_conn_is_type(conn, BT_CONN_TYPE_BR)) {
		LOG_DBG("Invalid connection type: %u for %p", conn->type, conn);
		return -EINVAL;
	}

	if (conn->state != BT_CONN_CONNECTED) {
		return -ENOTCONN;
	}
```

gzh-terry commented on subsys/bluetooth/host/conn.c:
Similar here, use `if (!bt_conn_is_type(conn, BT_CONN_TYPE_BR))`

gzh-terry commented on subsys/bluetooth/host/conn.c:
Remove extra line break
